### ISSUE TITLE
Tridactyl isn't loaded in pdf pages anymore

### ISSUE THREAD

**Issue by glacambre** [cmcaine/tridactyl#541]
PDF to reproduce this issue on: https://www.openacc.org/sites/default/files/inline-files/OpenACC.2.6.final.pdf
Tridactyl version: 1.12.0-11-g70ed5d1
Firefox version: 60.0

The console stays empty when using `set logging.messaging debug` and pressing random keys on pdf pages.

**Comment by glacambre** [comment]
Relevant bugzilla issue: https://bugzilla.mozilla.org/show_bug.cgi?id=1456485

**Comment by bovine3dom** [comment]
So it looks like it's a WONTFIX? How do we get a `mozillaAddons` permission :D?  Personally I think Tridactyl would be a smarter acquisition than Pocket, so I'll sell my share for a mere twice what Mozilla paid for that.

**Comment by glacambre** [comment]
Yup, probably WONTFIX. Maybe we could manage to inject Tridactyl everywhere with https://github.com/cmcaine/tridactyl/issues/460 ?

**Comment by bovine3dom** [comment]
Disabling the built in PDF viewer via user.js might be a better idea. In the grand tradition of the internet, it has awful font rendering anyway.

**Comment by glacambre** [comment]
This is fixed by `:fixamo`. Tridactyl's so great that even the wontfix are fixed :D 
Thanks @maxigaz !

**Comment by bovine3dom** [comment]
We could potentially work around this by just including the entire viewer in Tridactyl: https://github.com/mozilla/pdf.js/wiki/Setup-pdf.js-in-a-website

**Comment by nvkrishna77** [comment]
This is an important requirement. There are multiple instances where it impacts. While I can live without the tridactyl being present for browsing - there are other ways to do that. Similarly I goto other tabs - at least by using the mouse. But it doesn't let me find the url of the pdf - seems like there is *absolutely no way* to get this if my address bar has hidden.

**Comment by bovine3dom** [comment]
Yes. This is why `guiset gui none` only auto-hides the address bar.

**Comment by nvkrishna77** [comment]
Oh, that opens another big can of worms.

1. guiset gui none --> eats up the top lines of pages, when the navbar is hiding.
2. guiset gui none --> (whenever navbar is set to none) --> If I open a pdf or any page with tridactyl not active: I cannot switch to any other tab, without closing the current tab.

I cannot use the autohide or navbar = none.

**Comment by bovine3dom** [comment]
What's wrong with Ctrl tab?

I don't understand the eating of the top lines of pages. It works fine for me.

**Comment by nvkrishna77** [comment]
Ctrl-tab -- I had forgotten about that one. Thanks.

Top line: Here is a current page I am looking at
http://www.espncricinfo.com/series/8532/game/1153255/bangladesh-vs-india-final-asia-cup-2018-19

If you set gui none: you will see the missing top line.

**Comment by braham-snyder** [comment]
If I understand correctly (and I may not), it might help to vote on the following issue (you can even login with your github account), and it might also help to respond to their request for extension developer feedback: https://bugzilla.mozilla.org/show_bug.cgi?id=1454760#c27

> To get a better understanding of why extension developers want to run scripts in PDF pages, please share your use case if it is not covered in this bug or the linked duplicates. If you want to have a conversation, join #webextensions at irc.mozilla.org and ping me (IRC nick: robwu_nl), or send me an e-mail.

<br>

Again, I might be misunderstanding this, but it sounds like they might even accept a patch to fix this issue: https://bugzilla.mozilla.org/show_bug.cgi?id=1454760#c16

> > Could you, please, add a setting to the about:config page that allows access
> > to pdf.js content for browser extensions?
>
> I'd probably accept a patch for that.

**Comment by chrisvella** [comment]
Hi all I have a work around. I am using [google pdf viewer for firefox](https://github.com/def00111/google-pdf-viewer) because I want to sandbox pdf's from my browser. Not a perfect solution that does come at the cost of some privacy.

**Comment by aplaice** [comment]
(Obviously this issue is not a priority. I'm leaving this for reference purposes and in case it inspires somebody.)

> We could potentially work around this by just including the entire viewer in Tridactyl: https://github.com/mozilla/pdf.js/wiki/Setup-pdf.js-in-a-website

As a proof of feasibility, that's almost exactly what [Vimium C](https://chrome.google.com/webstore/detail/vimium-c/hfjbmagddngcpeloejdejnfgbamkjaeg/) does with its [PDF viewer](https://chrome.google.com/webstore/detail/pdf-viewer-for-vimium-c/nacjakoppgmdcpemlfnfegmlhipddanj/). (The very slight difference is that, as you can see, they created the PDF viewer as an additional extension, to avoid making the main one too heavy. Based on cursory reading of the "Support" tab it seems that the secondary extension communicates with the main one, so that a user's modification of the keybindings carries over from the main extension to the secondary one.) Obviously, since it's a chromium extension, they override chromium's PDFium with their copy of PDF.js, but overriding Firefox's PDF.js with an extension's version of PDF.js should work just as well.

I've only briefly tested it in a throw-away profile (note: you need both the main extension and the pdf viewer installed), but it seems to work pretty great — following links, searching and even caret/visual mode work. The main disadvantage is that a `chrome-extension://xxx` is prepended to all PDF URLs.

**Comment by bovine3dom** [comment]
https://github.com/gdh1995/pdf.js/commit/e6523aad658d7e75027c2d799b1e51caf7c01c3f this is where vimium-c is injecting stuff

I'm somewhat reluctant to fork it like that because PDFs are nasty, complicated and full of security issues, so I would want to keep it updated. That repo is about half a year out of date and it is already 1,000 commits behind master.

Maybe we can depend on it as a normal package - https://www.npmjs.com/package/pdfjs-dist
